### PR TITLE
[ricos-theme] ricos bundlesize optimization

### PR DESCRIPTION
**Pull request by oded-soffrin**
Due to the fact JSS (with jss-preset-default) increase our bundle w/ 12KB (gzipped) and it's a blocker for our consumers to adapt, i've moved theme strategy logic to a different package.

usage for those who wish to use it can be found in tests and storybook. 
(TPA needs to be adjusted in a different branch)

This is just a quickfix, we need to think about replacing JSS (or at least jss-preset-default) with a different, slimmer solution.

### CODE REVIEW

File: packages/ricos-common/web/src/RicosEngine.tsx [hunk 1]
```diff
@@ -71,12 +76,11 @@ export class RicosEngine extends Component<EngineProps> {
     );
 
     return {
-      strategyProps: merge(strategyProps, previewStrategyResult),
+      strategyProps: merge(strategiesProps, previewStrategyResult),
       previewContent,
-      rawCss,
+      html: [html],
```
Comment (mharsat) [inline packages/ricos-common/web/src/RicosEngine.tsx:81]: `html` is a bit vague. how about `styling || styles || styleTag || styleElement`?

File: packages/ricos-theme/web/src/themes/common.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /* eslint-disable camelcase */
-import { PaletteColors } from '../themeTypes';
+import { PaletteColors } from 'ricos-common/src/themeTypes';
```
Comment (mharsat) [inline packages/ricos-theme/web/src/themes/common.ts:2]:
```suggestion
import { PaletteColors } from 'ricos-common';
```

File: packages/ricos-theme/web/src/themes/editor-common.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /* eslint-disable max-len */
 import { adaptForeground, hexToRgbA, fallbackColor } from './utils';
-import { PaletteColors } from '../themeTypes';
+import { PaletteColors } from 'ricos-common/src/themeTypes';
```
Comment (mharsat) [inline packages/ricos-theme/web/src/themes/editor-common.ts:3]:
```suggestion
import { PaletteColors } from 'ricos-common';
```

File: packages/ricos-theme/web/src/themes/editor.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import { PaletteColors, ThemeUtils } from '../themeTypes';
+import { PaletteColors, ThemeUtils } from 'ricos-common/src/themeTypes';
```
Comment (mharsat) [inline packages/ricos-theme/web/src/themes/editor.ts:1]:
```suggestion
import { PaletteColors, ThemeUtils } from 'ricos-common';
```

File: packages/ricos-theme/web/src/themes/viewer.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /* eslint-disable camelcase */
-import { PaletteColors, ThemeUtils } from '../themeTypes';
+import { PaletteColors, ThemeUtils } from 'ricos-common/src/themeTypes';
```
Comment (mharsat) [inline packages/ricos-theme/web/src/themes/viewer.ts:2]:
```suggestion
import { PaletteColors, ThemeUtils } from 'ricos-common';
```

File: packages/ricos-theme/web/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "../../../tsconfig.json",
+  "compilerOptions": {
+    "declarationDir": "./dist",
+    "outDir": "./dist",
+    "allowJs": false
```
Comment (mharsat) [inline packages/ricos-theme/web/tsconfig.json:6]: default value, can be removed

**Review by mharsat — CHANGES_REQUESTED** [review]

File: packages/ricos-common/web/src/RicosTypes.ts [hunk 7]
```diff
@@ -67,17 +67,14 @@ export interface RicosEditorProps extends RicosProps {
...
 export interface RicosViewerProps extends RicosProps {
   plugins?: ViewerPluginConfig[];
   preview?: PreviewSettings;
   seoSettings?: boolean | SEOSettings;
-}
-
-export interface RicosTheme {
-  palette?: Palette | PalettePreset;
-  parentClass?: string;
+  createThemeStrategy?: () => ThemeStrategyFunction;
```
[the remark this answers is not in the document]
Comment (oded-soffrin) [reply packages/ricos-common/web/src/RicosTypes.ts:77]: agree on added once to RicosProps, but I thought palette & parenClass should be params supplied to themeStratgey.

File: examples/storybook/stories/Components/EditorWrapper.js [hunk 8]
```diff
@@ -161,8 +170,13 @@ class EditorWrapper extends React.Component {
         placeholder={'Share something...'}
         toolbarSettings={toolbarSettings}
         onChange={onChange}
+        createThemeStrategy={createThemeStrategy}
```
Comment (Barackos) [inline examples/storybook/stories/Components/EditorWrapper.js:173]: To simplify the API, I have a suggestion:
- no additional `createThemeStrategy` prop
- instead, `theme` prop API will look something like this:
`theme={createThemeStrategy({ palette, ... })`

- change RicosEngine's field inside the constructor to:
`this.themeStrategy = createThemeStrategy`

- modify the `createThemeStrategy` to accept those fields (palette, cssOverride, fonts..)

**Review by Barackos — APPROVED** [review]

File: packages/ricos-common/web/src/RicosEngine.tsx (hunk 1 above)
Comment (oded-soffrin) [reply packages/ricos-common/web/src/RicosEngine.tsx:81]: By inetention
I want to make it general. it was rawCss that we wrapped in <style> div. 
My goal was to make generic strategy function, that can add html (in this case styletag, but in other cases, who knows) to the rendering function

File: packages/ricos-theme/web/tsconfig.json (hunk 6 above)
Comment (oded-soffrin) [reply packages/ricos-theme/web/tsconfig.json:6]: no comparnede. it's defined in all tsconfig. can we remove from all ?

**Comment by mharsat** [comment]
> Due to the fact JSS (with jss-preset-default) increase our bundle w/ 12KB (gzipped) and it's a blocker for our consumers to adapt, i've moved theme strategy logic to a different package.
> 
> usage for those who wish to use it can be found in tests and storybook.
> (TPA needs to be adjusted in a different branch)
> 
> This is just a quickfix, we need to think about replacing JSS (or at least jss-preset-default) with a different, slimmer solution.

For replacing JSS we can consider having a presets package, that includes some function that creates custom themes. Something like
```jsx
import { bmPalette, createPalette } from 'ricos-theme';

<RicosEditor theme={{ palette: bmPalette }} />

<RicosEditor theme={{ palette: createPalette(args, that, create, palettes) }} />
```

File: packages/ricos-theme/web/tsconfig.json (hunk 6 above)
Comment (mharsat) [reply packages/ricos-theme/web/tsconfig.json:6]: Actually yes we can remove all instances of `"allowJs": false` :)

Comment (oded-soffrin) [reply packages/ricos-theme/web/tsconfig.json:6]: are you talking about 'allowJs' only ?

Comment (mharsat) [reply packages/ricos-theme/web/tsconfig.json:6]: yes, doesn't need to be in this PR

**Comment by oded-soffrin** [comment]
@Barackos @mharsat @lxgreen updated the API to be 
`theme={createThemeStrategy({ palette })}`

createThemeStrategy is a default export - any better name for examples ? 
maybe `createTheme` ?

other than that take a peek and if looks good i'll update the DOCS and storybook and will merge

**Comment by Barackos** [comment]
> @Barackos @mharsat @lxgreen updated the API to be
> `theme={createThemeStrategy({ palette })}`
> 
> createThemeStrategy is a default export - any better name for examples ?
> maybe `createTheme` ?
> 
> other than that take a peek and if looks good i'll update the DOCS and storybook and will merge

@oded-soffrin very cool! I think `createTheme` is a good name

**Comment by mharsat** [comment]
> @Barackos @mharsat @lxgreen updated the API to be
> `theme={createThemeStrategy({ palette })}`
> 
> createThemeStrategy is a default export - any better name for examples ?
> maybe `createTheme` ?
> 
> other than that take a peek and if looks good i'll update the DOCS and storybook and will merge

strategies is an intern implementation term, I also suggest `createTheme` as this is exactly what that does in consumer point of view.
also, I suggest named exports for APIs. we might want to export more things later and named exports provide more flexibility

other than that looks great, go ahead and update the docs

**Review by lxgreen — APPROVED** [review]

**Review by mharsat — APPROVED** [review]

**Pull request merged**